source/adios2/toolkit/burstbuffer/FileDrainer.cpp

```diff
@@ -0,0 +1,277 @@
+                 countBytes);
+}
+
+void FileDrainer::AddOperationWriteAt(const std::string &toFileName,
+                                      size_t toOffset, size_t countBytes,
+                                      const void *data)
+{
+    std::string emptyStr;
+    AddOperation(DrainOperation::WriteAt, emptyStr, toFileName, 0, toOffset,
+                 countBytes, data);
+}
+
+void FileDrainer::AddOperationWrite(const std::string &toFileName,
+                                    size_t countBytes, const void *data)
+{
+    std::string emptyStr;
+    AddOperation(DrainOperation::Write, emptyStr, toFileName, 0, 0, countBytes,
+                 data);
+}
+
+void FileDrainer::AddOperationOpen(const std::string &toFileName, Mode mode)
+{
+    std::string emptyStr;
+    if (mode == Mode::Write)
+    {
+        AddOperation(DrainOperation::Create, emptyStr, toFileName, 0, 0, 0);
+    }
+    else if (mode == Mode::Append)
+    {
+        AddOperation(DrainOperation::Open, emptyStr, toFileName, 0, 0, 0);
+    }
+    else
+    {
+        throw std::runtime_error(
+            "ADIOS Coding ERROR: FileDrainer::AddOperationOpen() only supports "
+            "Write and Append modes\n");
+    }
+}
+
+InputFile FileDrainer::GetFileForRead(const std::string &path)
+{
+    auto it = m_InputFileMap.find(path);
+    if (it != m_InputFileMap.end())
+    {
+        return it->second;
+    }
+    else
+    {
+        InputFile f = std::make_shared<std::ifstream>();
+        m_InputFileMap.emplace(path, f);
+        Open(f, path);
+        return f;
+    }
+}
+
+OutputFile FileDrainer::GetFileForWrite(const std::string &path, bool append)
+{
+    auto it = m_OutputFileMap.find(path);
+    if (it != m_OutputFileMap.end())
+    {
+        return it->second;
+    }
+    else
+    {
+        OutputFile f = std::make_shared<std::ofstream>();
+        m_OutputFileMap.emplace(path, f);
+        Open(f, path, append);
+        return f;
+    }
+}
+
+void FileDrainer::Open(InputFile &f, const std::string &path)
+{
+
+    f->rdbuf()->pubsetbuf(0, 0);
+    f->open(path, std::ios::in);
+}
+
+void FileDrainer::Open(OutputFile &f, const std::string &path, bool append)
+{
+
+    if (append)
+    {
+        f->rdbuf()->pubsetbuf(0, 0);
+        f->open(path, std::ios::out | std::ios::app | std::ios::binary);
+    }
+    else
+    {
+        f->rdbuf()->pubsetbuf(0, 0);
+        f->open(path, std::ios::out | std::ios::trunc | std::ios::binary);
+    }
+}
+
+void FileDrainer::Close(InputFile &f) { f->close(); }
+void FileDrainer::Close(OutputFile &f) { f->close(); }
+
+bool FileDrainer::Good(InputFile &f) { return (f->good()); }
+bool FileDrainer::Good(OutputFile &f) { return (f->good()); }
+
+void FileDrainer::CloseAll()
+{
+    for (auto it = m_OutputFileMap.begin(); it != m_OutputFileMap.end(); ++it)
+    {
+        // if (it->second->good())
+        //{
+        Close(it->second);
+        //}
+        m_OutputFileMap.erase(it);
+    }
```

Comment:
@pnorbert the test failures on Windows are due to invalidating the loop iterator here (and in the input loop below).  The `.erase()` removes the current iterator so the `++it` in end of the `for()` tries to increment an invalid iterator.  In a Debug build the MSVC iterator debugging catches it.  Otherwise this likely strays off and doesn't actually close or flush all of the files, hence the test failure.

I fixed it locally with

```patch
diff --git a/source/adios2/toolkit/burstbuffer/FileDrainer.cpp b/source/adios2/toolkit/burstbuffer/FileDrainer.cpp
index 72b0b85e..97eb793a 100644
--- a/source/adios2/toolkit/burstbuffer/FileDrainer.cpp
+++ b/source/adios2/toolkit/burstbuffer/FileDrainer.cpp
@@ -181,16 +181,16 @@ void FileDrainer::CloseAll()
         //{
         Close(it->second);
         //}
-        m_OutputFileMap.erase(it);
     }
+    m_OutputFileMap.clear();
     for (auto it = m_InputFileMap.begin(); it != m_InputFileMap.end(); ++it)
     {
         // if (it->second->good())
         //{
         Close(it->second);
         //}
-        m_InputFileMap.erase(it);
     }
+    m_InputFileMap.clear();
 }

 void FileDrainer::Seek(InputFile &f, size_t offset, const std::string &path)
```

With that the `Utils.IOTest.BurstBuffer.1to1.BP4.Write` test passes.

Reply:
Thanks for pointing out this mistake. After fixing this, the test passes on Windows only because the draining of metadata succeeds and the followup tests with bpls only read the metadata. The data, however, is not drained at all. 

```
Drain 0: Copy from bb\burstbuffer-32KB.bp\data.0 -> burstbuffer-32KB.bp\data.0 33025 bytes
ADIOS THREAD ERROR: FileDrainer couldn't read from file bb\burstbuffer-32KB.bp\data.0 offset = -1 count = 32927 bytes but only 0. File size at time of reading = -1
: iostream stream error
Drain 0: Copy from bb\burstbuffer-32KB.bp\md.0 -> burstbuffer-32KB.bp\md.0 277 bytes
Drain 0: Write to file burstbuffer-32KB.bp\md.idx 64 bytes of data from memory (no seek)
    App 1: Max write time = 0.0009631
    App 1: Min write time = 0.0009631
App 1 Step 3:
Drain 0: Copy from bb\burstbuffer-32KB.bp\data.0 -> burstbuffer-32KB.bp\data.0 32961 bytes  -- Skip because of previous error
Drain 0: Copy from bb\burstbuffer-32KB.bp\md.0 -> burstbuffer-32KB.bp\md.0 213 bytes
```

The current code skips further operations on any file where an error has happened, basically ruining the draining effort. What I am trying to figure out on Windows that how to recover from EOF status and read successfully later.  Resetting the flags with ifstream.clear() does not help, not even closing the file and opening a new one and starting from scratch. I wrote some standalone test where it works, just not here in this FileDrainer thread.

Reply:
The strange thing is that I can seek to the end-10 and read the last 10 bytes successfully from the file, but the actual read from offset 0 for the entire size of block (33025 bytes) fails. 

-- Read from bb\burstbuffer-32KB.bp\data.0 start = 0 count = 33025 fileSize = 33025 iostate = 0
-- Test read of 10 bytes from end resulted in iostate = 0 gcount = 10
-- Read reached EOF for bb\burstbuffer-32KB.bp\data.0. read = 98 bytes  iostate = 3

Reply:
@bradking Thank you for your suggestive question about binary mode. I forgot to add that flag for the read side. This was causing the problem I was seeing.